### PR TITLE
Clarify reason for setTextContent helper

### PR DESCRIPTION
**Pull request by jquense**
Some little thing’s I noticed, referencing the code here writing some of react-dom-lite

### CODE REVIEW

File: packages/react-dom/src/client/setTextContent.js [hunk 1]
```diff
@@ -1,36 +0,0 @@
...
-    if (
-      firstChild &&
-      firstChild === node.lastChild &&
-      firstChild.nodeType === TEXT_NODE
-    ) {
-      firstChild.nodeValue = text;
-      return;
-    }
-  }
-  node.textContent = text;
```
Comment (jquense) [inline packages/react-dom/src/client/setTextContent.js]: i'm not sure this is safe, but it seems to be. We use straight sets on textContent elsewhere

**Comment by jquense** [comment]
also, there _seems_ to be some duplication or extra work being done in `diffProperties`? I could definitely be reading it wrong, the loops there are hairy, but this loop: https://github.com/facebook/react/blob/8be5a8b7a38e347f338827f88bda98a0d34b05f0/packages/react-dom/src/client/ReactDOMFiberComponent.js#L639-L644

seems to be essentially duplicated lower down in the `nextProp` loop: [here](https://github.com/facebook/react/blob/8be5a8b7a38e347f338827f88bda98a0d34b05f0/packages/react-dom/src/client/ReactDOMFiberComponent.js#L690-L699)

File: packages/react-dom/src/client/ReactDOMFiberComponent.js [hunk 2]
```diff
@@ -292,12 +291,11 @@ function setInitialDOMProperties(
...
         // https://github.com/facebook/react/issues/6731#issuecomment-254874553
-        const canSetTextContent = tag !== 'textarea' || nextProp !== '';
-        if (canSetTextContent) {
-          setTextContent(domElement, nextProp);
+        if (tag !== 'textarea' || nextProp !== '') {
+          domElement.textContent = nextProp;
         }
       } else if (typeof nextProp === 'number') {
-        setTextContent(domElement, '' + nextProp);
+        domElement.textContent = '' + nextProp;
```
Comment (nhunzaker) [inline packages/react-dom/src/client/ReactDOMFiberComponent.js]: Why stringify here and not line 295?

File: packages/react-dom/src/client/ReactDOMFiberComponent.js [hunk 3]
```diff
@@ -340,7 +338,7 @@ function updateDOMProperties(
     } else if (propKey === DANGEROUSLY_SET_INNER_HTML) {
       setInnerHTML(domElement, propValue);
     } else if (propKey === CHILDREN) {
-      setTextContent(domElement, propValue);
+      domElement.textContent = propValue;
```
Comment (nhunzaker) [inline packages/react-dom/src/client/ReactDOMFiberComponent.js]: Likewise, should we stringify?

Comment (jquense) [reply packages/react-dom/src/client/ReactDOMFiberComponent.js]: I think the string if you happens in diffProperties for update, so the value is already a string here

Comment (nhunzaker) [reply packages/react-dom/src/client/ReactDOMFiberComponent.js]: Got it. In any case, this looks consistent with the old behavior. 👍

**Comment by jquense** [comment]
I'm not sure how to verify honestly. It's looks to me like the setTextContent was legacy considering the "modern" usages just use text Content, but perhaps someone with more perspective knows? In particular that one branch is really special, I was assuming it was for ie8 (per the comment) but I'm not sure the acceptance criteria for that

File: packages/react-dom/src/client/ReactDOMFiberComponent.js [hunk 4]
```diff
@@ -670,7 +668,7 @@ export function diffProperties(
   }
   for (propKey in nextProps) {
     const nextProp = nextProps[propKey];
-    const lastProp = lastProps != null ? lastProps[propKey] : undefined;
+    const lastProp = lastProps[propKey];
```
Comment (nhunzaker) [inline packages/react-dom/src/client/ReactDOMFiberComponent.js]: Just thinking about moving carefully, is this change related to the text content change? Could we send this out separately?

I know that's tip toeing quite a bit. Just trying to figure out how to move forward with confidence :)

Comment (jquense) [reply packages/react-dom/src/client/ReactDOMFiberComponent.js]: Good eye :P yeah it's not related to the one change, just another thing I noticed. I did split it out by commit but we can do separate PR's

No harm in being intentional about the changes!

**Comment by syranide** [comment]
@jquense `setTextContent` should be IE8 only.

**Comment by syranide** [comment]
@jquense **EDIT:** Specifically because `innerText` in IE8 behaves unacceptably and actually messes with the text being set.

**Comment by syranide** [comment]
@jquense Sorry for the spam, but here's my origin PR #1864 :)

**Comment by jquense** [comment]
Awesome, thanks for the context @syranide 

Anyone else wanna jump in? otherwise good to go?

**Comment by gaearon** [comment]
I don't think this code is currently IE8-motivated. The IE8 part was removed in https://github.com/facebook/react/pull/11331.

The current code using `nodeValue` was performance-motivated. @trueadm added it in #7005. If we can run benchmarks and confirm it's no longer causing a difference in modern browsers I think we can remove it, but sounds like we should at least verify that.

**Comment by jquense** [comment]
that's really interesting. The idea is that if the content is a single text node it's faster to set nodeValue there? I'd be happy try and verify, at the very least I'll update the comment :)

**Comment by trueadm** [comment]
It's still much faster to update text via `nodeValue` than `textContent`.

**Comment by jquense** [comment]
TIL :)

Thanks for the clarification (and sorry i didn't find the previous work on this before updating!)

**Comment by syranide** [comment]
@trueadm Just curious, could it possibly be faster with `.data` even? Considering that's a property of the Text-node itself rather then the Element.

**Comment by trueadm** [comment]
@syranide I tried with `data` but it wasn't as fast as `value` still.

File: packages/react-dom/src/client/ReactDOMFiberComponent.js (hunk 2 above)
Comment (trueadm) [reply packages/react-dom/src/client/ReactDOMFiberComponent.js]: `textContent` applies stringify and sanitization AFAIK

File: packages/react-dom/src/client/ReactDOM.js [hunk 5]
```diff
@@ -736,7 +737,7 @@ const DOMRenderer = ReactFiberReconciler({
     },
 
     resetTextContent(domElement: Instance): void {
-      domElement.textContent = '';
+      setTextContent(domElement, '');
```
Comment (nhunzaker) [inline packages/react-dom/src/client/ReactDOM.js]: Semi related:
https://github.com/facebook/react/issues/11609

 `commitTextUpdate` throws in some cases in IE9. I wonder if using this method there would fix it.

**Comment by trueadm** [comment]
I can also confirm `innerText` is faster than `textContent` for creating the initial text and also for updates.

**Comment by syranide** [comment]
@trueadm Just for posterity, you should absolutely not use `innerText` instead of `textContent` because they are not identical.

**Comment by trueadm** [comment]
@syranide I’m well aware. http://perfectionkills.com/the-poor-misunderstood-innerText/.

**Pull request merged**